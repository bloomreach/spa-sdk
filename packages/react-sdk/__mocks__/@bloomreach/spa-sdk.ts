/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const spaSdk = jest.requireActual('@bloomreach/spa-sdk');

export const META_POSITION_BEGIN = spaSdk.META_POSITION_BEGIN;
export const META_POSITION_END = spaSdk.META_POSITION_END;

export function mockMeta(data: string, position: string) {
  return {
    getData: jest.fn().mockReturnValue(data),
    getPosition: jest.fn().mockReturnValue(position),
  };
}

export function mockNoCommentMeta() {
  return mockMeta('not-a-comment', META_POSITION_BEGIN);
}

const componentMock = {
  getMeta: jest.fn().mockReturnValue([
    mockMeta('meta-begin', META_POSITION_BEGIN),
    mockMeta('meta-end', META_POSITION_END),
  ]),
};

export const pageMock = {
  getComponent: jest.fn().mockReturnValue(componentMock),
  getContent: jest.fn(),
  getTitle: jest.fn(),
  sync: jest.fn(),
};

export const initialize = jest.fn();
export const destroy = jest.fn();

export function isMetaComment(value: any) {
  return value.getData() !== 'not-a-comment';
}
