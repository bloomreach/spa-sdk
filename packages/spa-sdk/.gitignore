/coverage
/docs
/lib
/node_modules
.DS_store