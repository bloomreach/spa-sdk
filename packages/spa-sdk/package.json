{
  "name": "@bloomreach/spa-sdk",
  "version": "14.0.0",
  "description": "Bloomreach SPA SDK",
  "keywords": [
    "bloomreach",
    "spa",
    "sdk"
  ],
  "author": "Bloomreach B.V.",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://code.onehippo.org/cms-community/bloomreach-spa-sdk",
    "directory": "packages/spa-sdk"
  },
  "bugs": {
    "url": "https://issues.onehippo.com/projects/CMS/issues"
  },
  "main": "dist/index.js",
  "browser": "dist/index.js",
  "module": "dist/index.mjs",
  "es2015": "dist/index.es6.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist/"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js",
    "docs": "tsc --emitDeclarationOnly && dts-bundle-generator --out-file dist/spa-sdk.d.ts dist/index.d.ts && typedoc dist/spa-sdk.d.ts && minicat src/docs/typedoc.css >> docs/assets/css/main.css && rm dist/spa-sdk.d.ts",
    "lint": "tslint --project .",
    "test": "jest"
  },
  "jest": {
    "testEnvironment": "jsdom",
    "coverageDirectory": "./coverage",
    "collectCoverage": true,
    "collectCoverageFrom": [
      "src/**/*.ts"
    ],
    "moduleFileExtensions": [
      "ts",
      "js"
    ],
    "transform": {
      "^.+\\.ts$": "ts-jest"
    },
    "globals": {
      "ts-jest": {
        "tsConfig": "tsconfig.json"
      }
    },
    "testMatch": [
      "**/?(*.)spec.ts"
    ]
  },
  "devDependencies": {
    "@babel/core": "^7.6",
    "@babel/plugin-proposal-class-properties": "^7.5",
    "@babel/plugin-proposal-object-rest-spread": "^7.5",
    "@babel/plugin-transform-object-assign": "^7.2",
    "@babel/polyfill": "^7.6",
    "@babel/preset-env": "^7.6",
    "@types/jest": "^24.0",
    "@types/node": "^12.0",
    "babel-loader": "^8.0",
    "babel-plugin-transform-async-to-promises": "^0.8",
    "dts-bundle-generator": "^3.2",
    "jest": "^24.0",
    "jest-environment-jsdom-fifteen": "^1.0",
    "js-beautify": "^1.0",
    "minicat": "^1.0",
    "rollup": "^1.21",
    "rollup-plugin-babel": "^4.3",
    "rollup-plugin-terser": "^5.1",
    "rollup-plugin-typescript2": "^0.24",
    "ts-jest": "^24.0",
    "ts-loader": "^5.0",
    "tslint": "^5.0",
    "tslint-config-airbnb": "^5.0",
    "tslint-loader": "^3.5",
    "typedoc": "^0.15.0",
    "typescript": "^3.6"
  },
  "dependencies": {
    "emittery": "^0.4"
  }
}
