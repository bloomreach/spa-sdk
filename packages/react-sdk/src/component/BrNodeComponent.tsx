/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component } from '@bloomreach/spa-sdk';
import React, { useContext } from 'react';
import { BrMeta } from '../meta';
import { BrMappingContext } from './BrMappingContext';
import { BrProps } from './BrProps';

export function BrNodeComponent<T extends Component>(props: React.PropsWithChildren<BrProps<T>>): React.ReactElement {
  const context = useContext(BrMappingContext);
  const { component, children } = props;

  const mapping = component && (context[component.getName()] as React.ComponentType<BrProps>);
  const meta = component?.getMeta();

  const content = mapping ? React.createElement(mapping, props) : children;

  return React.createElement(BrMeta, { meta }, content);
}
