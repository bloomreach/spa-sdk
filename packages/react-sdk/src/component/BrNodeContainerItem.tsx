/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useCallback, useContext, useEffect, useState } from 'react';
import { ContainerItem, TYPE_CONTAINER_ITEM_UNDEFINED } from '@bloomreach/spa-sdk';
import { BrContainerItemUndefined } from '../cms';
import { BrProps } from './BrProps';
import { BrMappingContext } from './BrMappingContext';
import { BrMeta } from '../meta';

export function BrNodeContainerItem(
  props: React.PropsWithChildren<BrProps<ContainerItem>>,
): React.ReactElement {
  const context = useContext(BrMappingContext);
  const { component, page, children } = props;
  const [forcedRerenders, setForcedRerenders] = useState(0);

  const onUpdate = useCallback((): void => {
    // Trigger a rerender of this component due to a non-react 3rd party event
    setForcedRerenders((c) => c + 1);
  }, []);

  useEffect(() => {
    component?.on('update', onUpdate);

    return () => {
      component?.off('update', onUpdate);
    };
  }, [component, onUpdate]);

  useEffect(() => {
    // Skip the mount event page sync as BrPage just called it
    if (forcedRerenders > 0) {
      page?.sync();
    }
  }, [page, forcedRerenders]);

  const getMapping = (): React.ComponentType<BrProps> => {
    const type = component?.getType();

    if (type && type in context) {
      return context[type] as React.ComponentType<BrProps>;
    }

    return (
      (context[
        TYPE_CONTAINER_ITEM_UNDEFINED as any
      ] as React.ComponentType<BrProps>) ?? BrContainerItemUndefined
    );
  };

  const mapping = getMapping();
  const meta = component?.getMeta();
  const content = mapping ? React.createElement(mapping, props) : children;

  return React.createElement(BrMeta, { meta }, content);
}
