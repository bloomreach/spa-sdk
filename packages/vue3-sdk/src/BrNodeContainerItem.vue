<!--
  Copyright 2023 Bloomreach

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

    https://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  -->

<template>
  <Suspense>
    <component
      :is="mapping[componentType]"
      v-if="componentType && componentType in mapping"
      :component="component"
      :page="page"
    />

    <component
      :is="mapping[TYPE_CONTAINER_ITEM_UNDEFINED]"
      v-else-if="TYPE_CONTAINER_ITEM_UNDEFINED in mapping"
      :component="component"
      :page="page"
    />

    <br-container-item-undefined v-else-if="component" :component="component"/>
  </Suspense>
</template>

<script setup lang="ts">
import { component$, mapping$, page$ } from '@/providerKeys';
import type { ContainerItem } from '@bloomreach/spa-sdk';
import { TYPE_CONTAINER_ITEM_UNDEFINED } from '@bloomreach/spa-sdk';
import { computed, inject } from 'vue';
import BrContainerItemUndefined from './BrContainerItemUndefined.vue';

const page = inject(page$);
const mapping = inject(mapping$, {});
const component = inject<ContainerItem>(component$);
const componentType = computed(() => component?.getType());
</script>
