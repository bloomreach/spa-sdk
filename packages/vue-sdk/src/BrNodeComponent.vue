<!--
  Copyright 2020 Hippo B.V. (http://www.onehippo.com)

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  -->

<template>
  <br-meta :meta="component.getMeta()">
    <slot>
      <br-node-container-item v-if="isContainerItem(component)" />

      <div v-else-if="isContainer(component)">
        <br-node-component v-for="(component, key) in component.getChildren()" :key="key" :component="component" />
      </div>

      <br-node-component v-else v-for="(component, key) in component.getChildren()" :key="key" :component="component" />
    </slot>
  </br-meta>
</template>

<script lang="ts">
import { Component as SpaComponent, Page, isContainerItem, isContainer } from '@bloomreach/spa-sdk';
import { Component, InjectReactive, Prop, Provide, Vue } from 'vue-property-decorator';
import BrMeta from './BrMeta.vue';
import BrNodeContainerItem from './BrNodeContainerItem.vue';

@Component({
  components: {
    BrMeta,
    BrNodeContainerItem,
  },
  methods: {
    isContainerItem,
    isContainer,
  },
})
export default class BrNodeComponent extends Vue {
  @Prop() component!: SpaComponent;

  @InjectReactive() private page!: Page;

  @Provide() // ProvideReactive doesn't work with recursive components
  private component$() {
    return this.component;
  }
}
</script>
