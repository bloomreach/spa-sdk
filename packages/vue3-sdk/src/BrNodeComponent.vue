<!--
  - Copyright 2023 Bloomreach
  -
  - Licensed under the Apache License, Version 2.0 (the "License");
  - you may not use this file except in compliance with the License.
  - You may obtain a copy of the License at
  -
  -   http://www.apache.org/licenses/LICENSE-2.0
  -
  - Unless required by applicable law or agreed to in writing, software
  - distributed under the License is distributed on an "AS IS" BASIS,
  - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  - See the License for the specific language governing permissions and
  - limitations under the License.
  -->

<template>
  <br-meta :meta="meta">
    <slot>
      <br-node-container-item v-if="isContainerItem(component)"/>

      <br-node-container v-else-if="isContainer(component)">
        <br-node-component v-for="(component, key) in children" :key="key" :component="component"/>
      </br-node-container>

      <component v-else-if="name && name in mapping" :is="mapping[name]" :component="component" :page="page"/>

      <br-node-component v-else v-for="(component, key) in children" :key="key" :component="component"/>
    </slot>
  </br-meta>
</template>

<script setup lang="ts">
import BrMeta from '@/BrMeta.vue';
import BrNodeContainer from '@/BrNodeContainer.vue';
import BrNodeContainerItem from '@/BrNodeContainerItem.vue';
import { component$, mapping$, page$ } from '@/providerKeys';
import type { Component, Page } from '@bloomreach/spa-sdk';
import { isContainer, isContainerItem } from '@bloomreach/spa-sdk';
import { inject, provide } from 'vue';
import type { Ref } from 'vue';

const { component } = defineProps<{ component?: Component }>();
const page = inject(page$, {} as Ref<Page>);
const mapping = inject(mapping$, {});
const children = component?.getChildren();
const meta = component?.getMeta();
const name = component?.getName();

provide(component$, component);
</script>
