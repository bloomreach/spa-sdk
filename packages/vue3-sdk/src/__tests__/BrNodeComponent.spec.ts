/*
 * Copyright 2020-2023 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { Component, MetaCollection, Page } from '@bloomreach/spa-sdk';
import { isContainer, isContainerItem } from '@bloomreach/spa-sdk';
import { shallowMount } from '@vue/test-utils';
import type { GlobalMountOptions } from '@vue/test-utils/dist/types';
import type { Mocked } from 'vitest';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { defineComponent, nextTick, PropType, Ref, ref } from 'vue';
import type { BrMapping } from '../../typings';
import BrNodeComponent from '../BrNodeComponent.vue';
import BrNodeContainerItem from '../BrNodeContainerItem.vue';
import { mapping$, page$ } from '../providerKeys';

vi.mock('@bloomreach/spa-sdk');

describe('BrNodeComponent', () => {
  let component: Mocked<Component>;
  let mapping: Ref<BrMapping>;
  let page: Mocked<Page>;
  let global: GlobalMountOptions;

  const someComponentName = 'something';
  const SomeComponent = defineComponent({
    name: someComponentName,
    props: {
      page: {
        type: Object as PropType<Page>,
      },
      component: {
        type: Object as PropType<Component>,
      },
    },
  });

  beforeEach(() => {
    component = {
      getChildren: vi.fn(() => []),
      getMeta: vi.fn(),
      getName: vi.fn(() => someComponentName),
      getType: vi.fn(),
    } as unknown as typeof component;
    mapping = ref({ something: SomeComponent });
    page = {
      isPreview: vi.fn(),
    } as unknown as typeof page;

    global = {
      // Render out default slot fallback content
      renderStubDefaultSlot: true,
      provide: {
        [page$ as symbol]: ref(page),
        [mapping$ as symbol]: mapping,
      },
    };
  });

  describe('render', () => {
    it('should surround the component with meta-data', () => {
      component.getMeta.mockReturnValue(['a', 'b'] as unknown as MetaCollection);
      const wrapper = shallowMount(BrNodeComponent, { global, props: { component } });
      const metaComponent = wrapper.findComponent({ name: 'br-meta' });

      expect(metaComponent.exists()).toEqual(true);
    });

    it('should render a container item', async () => {
      vi.mocked(isContainerItem).mockReturnValueOnce(true);

      const wrapper = shallowMount(BrNodeComponent, { global, props: { component } });
      await nextTick();

      expect(wrapper.findComponent(BrNodeContainerItem).exists()).toBe(true);
    });

    it('should render a container', async () => {
      const child1 = { getName: () => 'child1', getId: () => 'child1' } as unknown as Component;
      const child2 = { getName: () => 'child2', getId: () => 'child2' } as unknown as Component;
      component.getChildren.mockReturnValue([child1, child2]);
      vi.mocked(isContainer).mockReturnValueOnce(true);

      const wrapper = shallowMount(BrNodeComponent, { global, props: { component } });
      await nextTick();

      const container = wrapper.findComponent({ name: 'br-node-container' });
      expect(container.exists()).toBe(true);
      const children = container.findAllComponents({ name: 'br-node-component' });
      expect(children.at(0)?.props().component.getName()).toBe('child1');
      expect(children.at(1)?.props().component.getName()).toBe('child2');
    });

    it('should render a mapped component', async () => {
      component.getName.mockReturnValue('something');

      const wrapper = shallowMount(BrNodeComponent, { global, props: { component } });
      await nextTick();

      const props = wrapper.findComponent(SomeComponent).props();

      expect(wrapper.html()).toMatchSnapshot();
      expect(props.component).toEqual(component);
      expect(props.page).toEqual(page);
    });

    it('should render children', async () => {
      const child1 = { getName: () => 'child1', getId: () => 'child1' } as unknown as Component;
      const child2 = { getName: () => 'child2', getId: () => 'child2' } as unknown as Component;
      component.getChildren.mockReturnValue([child1, child2]);

      mapping.value = {};
      const wrapper = shallowMount(BrNodeComponent, { global, props: { component } });
      await nextTick();

      const children = wrapper.findAllComponents({ name: 'br-node-component' });
      expect(wrapper.html()).toMatchSnapshot();
      expect(children.at(0)?.props().component.getName()).toBe('child1');
      expect(children.at(1)?.props().component.getName()).toBe('child2');
    });
  });
});
