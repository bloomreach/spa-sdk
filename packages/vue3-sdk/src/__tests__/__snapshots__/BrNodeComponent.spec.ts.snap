// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BrNodeComponent > render > should render a mapped component 1`] = `
"<br-meta-stub>
  <suspense-stub>
    <something-stub component=\\"[object Object]\\" page=\\"[object Object]\\"></something-stub>
  </suspense-stub>
</br-meta-stub>"
`;

exports[`BrNodeComponent > render > should render children 1`] = `
"<br-meta-stub>
  <suspense-stub>
    <br-node-component-stub component=\\"[object Object]\\"></br-node-component-stub>
    <br-node-component-stub component=\\"[object Object]\\"></br-node-component-stub>
  </suspense-stub>
</br-meta-stub>"
`;
